osquery/events/darwin/endpointsecurity.cpp

```diff
@@ -0,0 +1,232 @@
+
+REGISTER(EndpointSecurityPublisher, "event_publisher", "endpointsecurity")
+
+Status EndpointSecurityPublisher::setUp() {
+  if (__builtin_available(macos 10.15, *)) {
+    if (FLAGS_disable_endpointsecurity) {
+      return Status::failure(1,
+                             "EndpointSecurity is disabled via configuration");
+    }
+
+    auto handler = ^(es_client_t* client, const es_message_t* message) {
+      handleMessage(message);
+    };
+
+    auto result = es_new_client(&es_client_, handler);
+    switch (result) {
+    case ES_NEW_CLIENT_RESULT_SUCCESS: {
+      es_client_success_ = true;
+    }
+      return Status::success();
+    case ES_NEW_CLIENT_RESULT_ERR_INVALID_ARGUMENT:
+      return Status::failure(1, "invalid argument");
+    case ES_NEW_CLIENT_RESULT_ERR_INTERNAL:
+      return Status::failure(1, "EndpointSecurity client cannot communicate");
+    case ES_NEW_CLIENT_RESULT_ERR_NOT_ENTITLED:
+      return Status::failure(1, "EndpointSecurity client lacks entitlement");
+    case ES_NEW_CLIENT_RESULT_ERR_NOT_PERMITTED:
+      return Status::failure(
+          1, "EndpointSecurity client lacks user TCC permissions");
+    case ES_NEW_CLIENT_RESULT_ERR_NOT_PRIVILEGED:
+      return Status::failure(1,
+                             "EndpointSecurity client is not running as root");
+    case ES_NEW_CLIENT_RESULT_ERR_TOO_MANY_CLIENTS:
+      return Status::failure(
+          1, "Too many EndpointSecurity clients running on the system");
+    default:
+      return Status::failure(1, "EndpointSecurity: Unknown error");
+    }
+  } else {
+    return Status::failure(
+        1, "EndpointSecurity is only available on macOS 10.15 and higher");
+  }
+}
+
+void EndpointSecurityPublisher::configure() {
+  if (es_client_ == nullptr) {
+    return;
+  }
+
+  auto cache = es_clear_cache(es_client_);
+  if (cache != ES_CLEAR_CACHE_RESULT_SUCCESS) {
+    VLOG(1) << "Couldn't clear cache for EndpointSecurity client";
+    return;
+  }
+
+  for (auto& sub : subscriptions_) {
+    auto sc = getSubscriptionContext(sub->context);
+    auto events = sc->es_event_subscriptions_;
+    auto es_sub = es_subscribe(es_client_, &events[0], events.size());
+    if (es_sub != ES_RETURN_SUCCESS) {
+      VLOG(1) << "Couldn't subscribe to EndpointSecurity subsystem";
+    }
+  }
+}
+
+void EndpointSecurityPublisher::tearDown() {
+  if (es_client_ == nullptr) {
+    return;
+  }
+  es_unsubscribe_all(es_client_);
+
+  if (es_client_success_) {
+    es_delete_client(es_client_);
+    es_client_ = nullptr;
+  }
+}
+
+static inline std::string getPath(const es_process_t* p) {
+  return p->executable->path.length > 0 ? p->executable->path.data : "";
+}
+
+static inline std::string getSigningId(const es_process_t* p) {
+  return p->signing_id.length > 0 ? p->signing_id.data : "";
+}
+
+static inline std::string getTeamId(const es_process_t* p) {
+  return p->team_id.length > 0 ? p->team_id.data : "";
```

Comment:
I see you are checking that `t->data != nullptr` for `getStringFromToken`, do you also need to check for null pointers in `getTeamId` and so on?

Reply:
Good catch, thanks!